ConfidentialClientTokenCache/Program.cs

```diff
@@ -0,0 +1,189 @@
+using Microsoft.Identity.Web.TokenCacheProviders.Distributed;
+using Microsoft.Identity.Web.TokenCacheProviders.InMemory;
+using System;
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace ConfidentialClientTokenCache
+{
+    /// <summary>
+    /// Choice of token cache serialization implementation
+    /// (this is for the sample)
+    /// </summary>
+    public enum CacheImplementationDemo
+    {
+        InMemory,
+        DistributedMemory,
+        DistributedSqlServer,
+        StackExchangeRedis,
+        CosmosDb
+    }
+
+    /// <summary>
+    /// The goal of this little sample is to show how you can use MSAL token cache adapters
+    /// in confidential client applications both in .NET Core, or .NET Framework 4.7.2
+    /// Note that if you write a .NET Core web app or web API, Microsoft recommends that you use the
+    /// IDownstreamApi or ITokenAcquisition interfaces directly (instead of MSAL)
+    /// </summary>
+    static class Program
+    {
+        static async Task Main(string[] args)
+        {
+            string clientId = "812287fd-3ea3-49c6-b4ab-e8d41dea1f17";
+            string clientSecret = "[Enter here the tenantID or domain name for your Azure AD tenant]";
+            string tenant = "msidentitysamplestesting.onmicrosoft.com";
+            string[] scopes = new[] { "api://2d96f90e-a1a7-4ef5-b15c-87758986eb1a/.default" };
+
+            CacheImplementationDemo cacheImplementation = CacheImplementationDemo.InMemory;
+
+            // Create the token cache (4 possible implementations)
+            IMsalTokenCacheProvider msalTokenCacheProvider = CreateTokenCache(cacheImplementation);
+
+            // Create the confidential client application
+            IConfidentialClientApplication app;
+            app = ConfidentialClientApplicationBuilder.Create(clientId)
+                .WithClientSecret(clientSecret)
+                .WithTenantId(tenant)
+                .Build();
+
+            await msalTokenCacheProvider.InitializeAsync(app.UserTokenCache);
+            await msalTokenCacheProvider.InitializeAsync(app.AppTokenCache);
+
+            // Acquire a token (twice)
+            var result = await app.AcquireTokenForClient(scopes)
+                .ExecuteAsync();
+            Console.WriteLine(result.AuthenticationResultMetadata.TokenSource);
+
+            result = await app.AcquireTokenForClient(scopes)
+                .ExecuteAsync();
+            Console.WriteLine(result.AuthenticationResultMetadata.TokenSource);
+
+        }
+
+        /// <summary>
+        /// Creates a token cache (implementation of your choice)
+        /// </summary>
+        /// <param name="cacheImplementation">implementation for the token cache</param>
+        /// <returns>An Msal Token cache provider</returns>
+        private static IMsalTokenCacheProvider CreateTokenCache(CacheImplementationDemo cacheImplementation=CacheImplementationDemo.InMemory)
```

Comment:
wonder how much of this can be moved into one of the libraries? make it even simpler.

Reply:
Awesome idea. Let's think what we can do for the options.